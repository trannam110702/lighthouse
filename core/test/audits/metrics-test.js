/**
 * @license
 * Copyright 2018 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import jestMock from 'jest-mock';

import MetricsAudit from '../../audits/metrics.js';
import {Interactive} from '../../computed/metrics/interactive.js';
import {getURLArtifactFromDevtoolsLog, readJson} from '../test-utils.js';
import {defaultSettings} from '../../config/constants.js';

const pwaTrace = readJson('../fixtures/artifacts/progressive-app/trace.json', import.meta);
const pwaDevtoolsLog = readJson('../fixtures/artifacts/progressive-app/devtoolslog.json', import.meta);
const lcpTrace = readJson('../fixtures/artifacts/paul/trace.json', import.meta);
const lcpDevtoolsLog = readJson('../fixtures/artifacts/paul/devtoolslog.json', import.meta);
const lcpImageTrace = readJson('../fixtures/traces/amp-m86.trace.json', import.meta);
const lcpImageDevtoolsLog = readJson('../fixtures/traces/amp-m86.devtoolslog.json', import.meta);
const lcpAllFramesTrace = readJson('../fixtures/traces/frame-metrics-m89.json', import.meta);
const lcpAllFramesDevtoolsLog = readJson('../fixtures/traces/frame-metrics-m89.devtools.log.json', import.meta);
const clsAllFramesTrace = readJson('../fixtures/traces/frame-metrics-m90.json', import.meta);
const clsAllFramesDevtoolsLog = readJson('../fixtures/traces/frame-metrics-m90.devtools.log.json', import.meta);
const jumpyClsTrace = readJson('../fixtures/traces/jumpy-cls-m90.json', import.meta);
const jumpyClsDevtoolsLog = readJson('../fixtures/traces/jumpy-cls-m90.devtoolslog.json', import.meta);

const settings = JSON.parse(JSON.stringify(defaultSettings));

describe('Performance: metrics', () => {
  // TODO(15841): investigate failures
  if (process.env.INTERNAL_LANTERN_USE_TRACE !== undefined) {
    return;
  }

  it('evaluates valid input correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input correctly (throttlingMethod=provided)', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'provided'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp from all frames) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpAllFramesDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'provided'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with image lcp) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpImageDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpImageTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpImageDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('leaves CLS undefined in an old trace without weighted scores', async () => {
    const artifacts = {
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: undefined,
      observedCumulativeLayoutShift: undefined,
    });
  });

  it('evaluates CLS correctly across all frames', async () => {
    const URL = getURLArtifactFromDevtoolsLog(clsAllFramesDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: clsAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: clsAllFramesDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'provided'},
      computedCache: new Map(),
    };
    const {details} = await MetricsAudit.audit(artifacts, context);

    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: expect.toBeApproximately(0.026463, 6),
      observedCumulativeLayoutShift: expect.toBeApproximately(0.026463, 6),
    });
  });

  it('does not fail the entire audit when TTI errors', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const mockTTIFn = jestMock.spyOn(Interactive, 'request');
    mockTTIFn.mockRejectedValueOnce(new Error('TTI failed'));
    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0].interactive).toEqual(undefined);
  });

  it('evaluates CLS correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(jumpyClsDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: jumpyClsTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: jumpyClsDevtoolsLog,
      },
    };

    const context = {
      settings: {...settings, throttlingMethod: 'simulate'},
      computedCache: new Map(),
    };
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: expect.toBeApproximately(2.268816, 6),
      observedCumulativeLayoutShift: expect.toBeApproximately(2.268816, 6),
    });
  });
});
