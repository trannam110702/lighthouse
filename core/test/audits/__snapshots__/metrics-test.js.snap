// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance: metrics evaluates valid input (with image lcp) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftMainFrame": 0,
  "firstContentfulPaint": 3364,
  "firstContentfulPaintAllFrames": undefined,
  "firstContentfulPaintAllFramesTs": undefined,
  "firstContentfulPaintTs": undefined,
  "interactive": 6354,
  "interactiveTs": undefined,
  "largestContentfulPaint": 5567,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "lcpLoadEnd": 5519,
  "lcpLoadStart": 5397,
  "maxPotentialFID": 160,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftMainFrame": 0,
  "observedDomContentLoaded": 2474,
  "observedDomContentLoadedTs": 760623117637,
  "observedFirstContentfulPaint": 2685,
  "observedFirstContentfulPaintAllFrames": 2685,
  "observedFirstContentfulPaintAllFramesTs": 760623328111,
  "observedFirstContentfulPaintTs": 760623328111,
  "observedFirstPaint": 2685,
  "observedFirstPaintTs": 760623328111,
  "observedFirstVisualChange": 2652,
  "observedFirstVisualChangeTs": 760623295599,
  "observedLargestContentfulPaint": 2882,
  "observedLargestContentfulPaintAllFrames": 2882,
  "observedLargestContentfulPaintAllFramesTs": 760623525330,
  "observedLargestContentfulPaintTs": 760623525330,
  "observedLastVisualChange": 3269,
  "observedLastVisualChangeTs": 760623912599,
  "observedLoad": 2697,
  "observedLoadTs": 760623340349,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 760620643599,
  "observedSpeedIndex": 2795,
  "observedSpeedIndexTs": 760623438456,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 760620643599,
  "observedTraceEnd": 4778,
  "observedTraceEndTs": 760625421283,
  "speedIndex": 5546,
  "speedIndexTs": undefined,
  "timeToFirstByte": 2394,
  "timeToFirstByteTs": undefined,
  "totalBlockingTime": 162,
}
`;

exports[`Performance: metrics evaluates valid input (with lcp from all frames) correctly 1`] = `
Object {
  "cumulativeLayoutShift": undefined,
  "cumulativeLayoutShiftMainFrame": undefined,
  "firstContentfulPaint": 863,
  "firstContentfulPaintAllFrames": 683,
  "firstContentfulPaintAllFramesTs": 23466705983,
  "firstContentfulPaintTs": 23466886143,
  "interactive": 863,
  "interactiveTs": 23466886143,
  "largestContentfulPaint": 863,
  "largestContentfulPaintAllFrames": 683,
  "largestContentfulPaintAllFramesTs": 23466705983,
  "largestContentfulPaintTs": 23466886143,
  "lcpLoadEnd": undefined,
  "lcpLoadStart": undefined,
  "maxPotentialFID": 16,
  "observedCumulativeLayoutShift": undefined,
  "observedCumulativeLayoutShiftMainFrame": undefined,
  "observedDomContentLoaded": 596,
  "observedDomContentLoadedTs": 23466619325,
  "observedFirstContentfulPaint": 863,
  "observedFirstContentfulPaintAllFrames": 683,
  "observedFirstContentfulPaintAllFramesTs": 23466705983,
  "observedFirstContentfulPaintTs": 23466886143,
  "observedFirstPaint": 616,
  "observedFirstPaintTs": 23466639588,
  "observedFirstVisualChange": 609,
  "observedFirstVisualChangeTs": 23466632130,
  "observedLargestContentfulPaint": 863,
  "observedLargestContentfulPaintAllFrames": 683,
  "observedLargestContentfulPaintAllFramesTs": 23466705983,
  "observedLargestContentfulPaintTs": 23466886143,
  "observedLastVisualChange": 5881,
  "observedLastVisualChangeTs": 23471904130,
  "observedLoad": 673,
  "observedLoadTs": 23466696096,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 23466023130,
  "observedSpeedIndex": 1583,
  "observedSpeedIndexTs": 23467605703,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 23466023130,
  "observedTraceEnd": 6006,
  "observedTraceEndTs": 23472029453,
  "speedIndex": 1583,
  "speedIndexTs": 23467606130,
  "timeToFirstByte": 565,
  "timeToFirstByteTs": 23466588051,
  "totalBlockingTime": 0,
}
`;

exports[`Performance: metrics evaluates valid input (with lcp) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0.23332411101386594,
  "cumulativeLayoutShiftMainFrame": 0.23332411101386594,
  "firstContentfulPaint": 1059,
  "firstContentfulPaintAllFrames": undefined,
  "firstContentfulPaintAllFramesTs": undefined,
  "firstContentfulPaintTs": undefined,
  "interactive": 3149,
  "interactiveTs": undefined,
  "largestContentfulPaint": 1524,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "lcpLoadEnd": undefined,
  "lcpLoadStart": undefined,
  "maxPotentialFID": 163,
  "observedCumulativeLayoutShift": 0.23332411101386594,
  "observedCumulativeLayoutShiftMainFrame": 0.23332411101386594,
  "observedDomContentLoaded": 331,
  "observedDomContentLoadedTs": 343577515428,
  "observedFirstContentfulPaint": 292,
  "observedFirstContentfulPaintAllFrames": 292,
  "observedFirstContentfulPaintAllFramesTs": 343577475882,
  "observedFirstContentfulPaintTs": 343577475882,
  "observedFirstPaint": 292,
  "observedFirstPaintTs": 343577475882,
  "observedFirstVisualChange": 207,
  "observedFirstVisualChangeTs": 343577391048,
  "observedLargestContentfulPaint": 292,
  "observedLargestContentfulPaintAllFrames": 292,
  "observedLargestContentfulPaintAllFramesTs": 343577475882,
  "observedLargestContentfulPaintTs": 343577475882,
  "observedLastVisualChange": 391,
  "observedLastVisualChangeTs": 343577575048,
  "observedLoad": 436,
  "observedLoadTs": 343577620312,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 343577184048,
  "observedSpeedIndex": 261,
  "observedSpeedIndexTs": 343577445114,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 343577184048,
  "observedTraceEnd": 3479,
  "observedTraceEndTs": 343580662857,
  "speedIndex": 1059,
  "speedIndexTs": undefined,
  "timeToFirstByte": 638,
  "timeToFirstByteTs": undefined,
  "totalBlockingTime": 84,
}
`;

exports[`Performance: metrics evaluates valid input correctly (throttlingMethod=provided) 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftMainFrame": 0,
  "firstContentfulPaint": 192,
  "firstContentfulPaintAllFrames": 192,
  "firstContentfulPaintAllFramesTs": 376406173872,
  "firstContentfulPaintTs": 376406173872,
  "interactive": 224,
  "interactiveTs": 376406205074,
  "largestContentfulPaint": 229,
  "largestContentfulPaintAllFrames": 229,
  "largestContentfulPaintAllFramesTs": 376406210378,
  "largestContentfulPaintTs": 376406210378,
  "lcpLoadEnd": 199,
  "lcpLoadStart": 198,
  "maxPotentialFID": 16,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftMainFrame": 0,
  "observedDomContentLoaded": 224,
  "observedDomContentLoadedTs": 376406205074,
  "observedFirstContentfulPaint": 192,
  "observedFirstContentfulPaintAllFrames": 192,
  "observedFirstContentfulPaintAllFramesTs": 376406173872,
  "observedFirstContentfulPaintTs": 376406173872,
  "observedFirstPaint": 192,
  "observedFirstPaintTs": 376406173872,
  "observedFirstVisualChange": 194,
  "observedFirstVisualChangeTs": 376406175564,
  "observedLargestContentfulPaint": 229,
  "observedLargestContentfulPaintAllFrames": 229,
  "observedLargestContentfulPaintAllFramesTs": 376406210378,
  "observedLargestContentfulPaintTs": 376406210378,
  "observedLastVisualChange": 4827,
  "observedLastVisualChangeTs": 376410808564,
  "observedLoad": 245,
  "observedLoadTs": 376406227054,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 376405981564,
  "observedSpeedIndex": 379,
  "observedSpeedIndexTs": 376406360609,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 376405981564,
  "observedTraceEnd": 11305,
  "observedTraceEndTs": 376417286238,
  "speedIndex": 379,
  "speedIndexTs": 376406360564,
  "timeToFirstByte": 170,
  "timeToFirstByteTs": 376406151592,
  "totalBlockingTime": 0,
}
`;

exports[`Performance: metrics evaluates valid input correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftMainFrame": 0,
  "firstContentfulPaint": 1107,
  "firstContentfulPaintAllFrames": undefined,
  "firstContentfulPaintAllFramesTs": undefined,
  "firstContentfulPaintTs": undefined,
  "interactive": 1122,
  "interactiveTs": undefined,
  "largestContentfulPaint": 1120,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "lcpLoadEnd": 972,
  "lcpLoadStart": 971,
  "maxPotentialFID": 50,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftMainFrame": 0,
  "observedDomContentLoaded": 224,
  "observedDomContentLoadedTs": 376406205074,
  "observedFirstContentfulPaint": 192,
  "observedFirstContentfulPaintAllFrames": 192,
  "observedFirstContentfulPaintAllFramesTs": 376406173872,
  "observedFirstContentfulPaintTs": 376406173872,
  "observedFirstPaint": 192,
  "observedFirstPaintTs": 376406173872,
  "observedFirstVisualChange": 194,
  "observedFirstVisualChangeTs": 376406175564,
  "observedLargestContentfulPaint": 229,
  "observedLargestContentfulPaintAllFrames": 229,
  "observedLargestContentfulPaintAllFramesTs": 376406210378,
  "observedLargestContentfulPaintTs": 376406210378,
  "observedLastVisualChange": 4827,
  "observedLastVisualChangeTs": 376410808564,
  "observedLoad": 245,
  "observedLoadTs": 376406227054,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 376405981564,
  "observedSpeedIndex": 379,
  "observedSpeedIndexTs": 376406360609,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 376405981564,
  "observedTraceEnd": 11305,
  "observedTraceEndTs": 376417286238,
  "speedIndex": 1107,
  "speedIndexTs": undefined,
  "timeToFirstByte": 600,
  "timeToFirstByteTs": undefined,
  "totalBlockingTime": 0,
}
`;
